packages/mui-joy/src/Typography/Typography.tsx

```diff
@@ -0,0 +1,275 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import clsx from 'clsx';
+import { OverridableComponent } from '@mui/types';
+import { unstable_extendSxProp as extendSxProp } from '@mui/system';
+import { unstable_composeClasses as composeClasses } from '@mui/base';
+import { unstable_capitalize as capitalize } from '@mui/utils';
+import { TypographyTypeMap, TypographyProps } from './TypographyProps';
+import styled from '../styles/styled';
+import useThemeProps from '../styles/useThemeProps';
+import { getTypographyUtilityClass } from './typographyClasses';
+
+const useUtilityClasses = (ownerState: TypographyProps) => {
+  const { align, gutterBottom, noWrap, paragraph, variant, level } = ownerState;
+
+  const slots = {
+    root: [
+      'root',
+      !level && variant,
+      level,
+      align && align !== 'inherit' && `align${capitalize(align)}`,
+      gutterBottom && 'gutterBottom',
+      noWrap && 'noWrap',
+      paragraph && 'paragraph',
+    ],
+  };
+
+  return composeClasses(slots, getTypographyUtilityClass, {});
+};
+
+export const TypographyRoot = styled('span', {
+  name: 'MuiTypography',
+  slot: 'Root',
+  overridesResolver: (props, styles) => {
+    const { ownerState } = props;
+
+    return [
+      styles.root,
+      ownerState.variant && styles[ownerState.variant],
+      ownerState.level && styles[ownerState.level],
+      ownerState.align !== 'inherit' && styles[`align${capitalize(ownerState.align)}`],
+      ownerState.noWrap && styles.noWrap,
+      ownerState.gutterBottom && styles.gutterBottom,
+      ownerState.paragraph && styles.paragraph,
+    ];
+  },
+})<{ ownerState: TypographyProps }>(({ theme, ownerState }) => ({
+  margin: 0,
+  // TODO: remove in the next major version
+  ...(ownerState.variant &&
+    ownerState.variant !== 'inherit' &&
+    theme.typography[ownerState.variant]),
+  // TODO: remove in the next major version
+  ...(ownerState.align !== 'inherit' && {
+    textAlign: ownerState.align,
+  }),
+  ...(ownerState.level && ownerState.level !== 'inherit' && theme.typography[ownerState.level]),
+  ...(ownerState.noWrap && {
+    overflow: 'hidden',
+    textOverflow: 'ellipsis',
+    whiteSpace: 'nowrap',
+  }),
+  ...(ownerState.gutterBottom && {
+    marginBottom: '0.35em',
+  }),
+  ...(ownerState.paragraph && {
+    marginBottom: 16,
+  }),
+}));
+
+const defaultVariantMapping = {
+  h1: 'h1',
+  h2: 'h2',
+  h3: 'h3',
+  h4: 'h4',
+  h5: 'h5',
+  h6: 'h6',
+  body1: 'p',
+  body2: 'p',
+  body3: 'p',
+  inherit: 'p',
+};
+
+const Typography = React.forwardRef(function Typography(inProps, ref) {
```

Comment:
Side note: I'm wondering why don't we just call it `Text` - it's much shorter to type :)

Reply:
I don't want to change the component name at this point but I think @oliviertassinari can explain the history.